File: leetcode/114-二叉树展开为链表/114-二叉树展开为链表.go
package main

/**
 * Definition for a binary tree node.

 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 114-二叉树展开为链表
func flatten(root *TreeNode) {
	if root == nil {
		return
	}
	flatten(root.Left)
	flatten(root.Right)
	// 1、左右⼦树已经被拉平成⼀条链表
	left := root.Left
	right := root.Right

	// 2、将左⼦树作为右⼦树
	root.Right = left
	root.Left = nil

	// 3、将原先的右⼦树接到当前右⼦树的末端
	// 寻找原本左子树的最右结点，将原本二叉树的右子树连接到原本二叉树左子树的最右结点上
	p := root
	for root.Right != nil {
		p = p.Right
	}
	p.Right = right
}

File: leetcode/114-二叉树展开为链表/114-二叉树展开为链表.js
/**
 * Definition for a binary tree node.
 * 
 */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */

// 114-二叉树展开为链表
var flatten = function (root) {
    if (!root) { 
        return;
    }
    flatten(root.left);
    flatten(root.right);

    const left = root.left;
    const right = root.right;
    root.right = left;
    root.left = null;

    const cur = root;
    while (root.right) {
        cur = cur.right;
    }
    cur.right = right;
};